### ISSUE TITLE
Unnecessary leading semi

### ISSUE THREAD

**Issue by wenfangdu** [prettier/prettier#9060]
**Prettier 2.1.0**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAeeBbADgGwIbwB8AOlAARmoBuAtBnlmQAI4R4Amc7AvMSABT1G3MgBI4VBDACUZAGRyyASyhKYAWQb9pfQqgD0tISSgHMuAnBMgANCAhYYS6AGdkoPACdPEAO4AFLwQ3FDwqCCV2WxAAI088MABrOBgAZSwElQBzZBhPAFc4OzgMGK5OdgAZPCgs-LwsuAAxCE96GCda5BA8fJgIaIALGAwcAHVBtTgXDLA4VOC1JSo1AE9usBc3OxUXOE8Yf3is+mQAMzwcPbsAKxcADwAheKSU1LwMOEqVOHPL65Ad3uqWyODgAEV8hB4H8rkUQBlPHtPN0qIVolhPCoYGNIjBBsgABwABjsmIgezG8Sw3Ux032kmiAEcofAjg4Qj0XDQoHAuFxop44CylEKjg1TkgLnC7HsMEpYQCXKCIazflL-vCYHgYrj2PjkAAmOx5PBKHDZADCEAwkpAUGgvzs+T2ABUdSFpQC0XAAJJQTiwVJgLGOACCAdSMFWYMVcAAvvGgA)
```sh
--arrow-parens avoid
--parser vue
--no-semi
--trailing-comma none
```

**Input:**
```vue
<template>
  <v-map @loaded="(map = $event) && initMap()"></v-map>
</template>

```

**Output:**
```vue
<template>
  <v-map @loaded=";(map = $event) && initMap()"></v-map>
</template>

```

**Expected behavior:**
```vue
<template>
  <v-map @loaded="(map = $event) && initMap()"></v-map>
</template>
```

**Comment by fisker** [comment]
This is expected and we format js like this, when `--no-semi`, but looks weird in this case.

**Comment by wenfangdu** [comment]
Yes, when code like `;(() => {...})()` is placed at start of js file, the leading semi is unnecessary too, can this behavior be adjusted?

**Comment by fisker** [comment]
We won't change current behavior in js files, but this one may worth discuss.

**Comment by brodybits** [comment]
Here is a case where I think the automatic semicolon insertion should be reconsidered:

**Prettier 2.1.0**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBLAZgAgBQEMCUmwAOlJpgEbb6kC+ppeAzgJ5RjWYC8AfEaeVwB3XKhiZc1ARJFjKUqLXzUQAGhAQADjFTQmyULgBORiEIAKxhPpS4ANiJb71FI7jABrODADKm96hQAObIMEYArnDqcAC2FHAAJgmJADK4weG4QXAAYhBGMbgwOsHIILjhMBBqIAAWMDF2AOq1YnBM-mBwPtZiqABuYixlYEzOaFBMcEYw5m5Bhcjo9lPqAFZMAB4AQm6e3j64MXApgXBLK1EgG5s+gUF2cACK4RDwF3arIP5GU0ZlFFw8TsNU0RkCMCaqASMFqyAAHAAGdRgiBTJpuTRlMHtab9c7qACOr3gcy0NnKTAAtFA4IlEjUjHBiagmXMsoskMtPlcpjFUB8vkx7o8Xm9zlzLuoYECoTC4UgAEzStyoOz3ADCEBinJA7QArDVwlMACpAmzcr79SIASSgyVgPjA4O0AEF7T4YCxHoK4LRaEA)
```sh
--parser babel
--no-semi
```

**Input:**
```jsx
if (a) {
  b()
}

(async() => {
  await a()
  await b()
})()
```

**Output:**
```jsx
if (a) {
  b()
}

;(async () => {
  await a()
  await b()
})()

```

**Comment by fisker** [comment]
> maybe we need to protect against the programmer adding code before the expression statement in the future

Yes, we won't change it in js files. The behavior for js  file  is  out scope of  this issue.

**Comment by brodybits** [comment]
> We won't change current behavior in js files

Here is the rationale, linked from the `--no-semi` option: https://prettier.io/docs/en/rationale.html#semicolons

(my apologies for the other comment)

> this one may worth discuss

It would definitely be nice to be consistent with Angular & HTML parsers, which would format that template unchanged.

**Comment by zhouyu1993** [comment]
Input:

`
[1, 2].forEach((item) => console.log(item))

[1, 2].forEach((item) => console.log(item))
`

Output:

`
;[1, 2]
  .forEach((item) => console.log(item))

  [(1, 2)].forEach((item) => console.log(item))
`

The first line, I can understand.

The second line, I'm shocked.

**Comment by fisker** [comment]
Run 

```js
console.log(
[1, 2].forEach((item) => console.log(item))

[1, 2].forEach((item) => console.log(item))
)
```

to understand it.

**Comment by zhouyu1993** [comment]
> Run
> 
> ```js
> console.log(
> [1, 2].forEach((item) => console.log(item))
> 
> [1, 2].forEach((item) => console.log(item))
> )
> ```
> 
> and
> 
> ```js
> [1, 2].map((item) => item)
> 
> [0, 1].toFixed(10)
> ```
> 
> to understand it.

Why is it wrong

**Comment by fisker** [comment]
If you still can't get it, parser your code on https://astexplorer.net/, and check the AST.

**Comment by eevleevs** [comment]
Please reconsider, an expression with leading semicolon is invalid in Alpine.js. Currently need to ignore every such node.

**Comment by zhouyu1993** [comment]
已收到。谢谢，我会尽快回复你。

**Comment by fisker** [comment]
Fixed by #14557

**Comment by zhouyu1993** [comment]
已收到。谢谢，我会尽快回复你。
